### ISSUE TITLE
The database has to be prepared prior to calling get_ce_calc

### ISSUE THREAD

**Issue by phymalidoust** [davidkleiven/CEMC#65]
Hi,

When do we get this error? " The database has to be prepared prior to calling get_ce_calc "

This error arises when I call 'get_ce_calc' in 2 processors or more. But when I call it by a single processor it passes without raising this error.

**Comment by davidkleiven** [comment]
It is because if the DB does not exist, only one processor can write. The
problem is that the write statement is so slow that you get a race
condition. 1 processor start to write, the others see that the entry is
there and tries to read. But since the write is not finished the read
fails. It is the CE code that creates the DB so it needs to be handled
there. As of now you don't get any speedup by running that part in
parallel, but maybe in the future.

Den fre. 11. jan. 2019, 11.11 skrev phymalidoust <notifications@github.com:

**Comment by phymalidoust** [comment]
I first thought that exactly what you say is happening. Then forced the processors to pop-up one after another and take their own copies. However, it stucks at the master processor.
Even the first processor creates its own DB etc but stucks in this function "get_ce_calc" and doesn't allow the other ones to pop-up.

**Comment by phymalidoust** [comment]
This routine "get_ce_calc" prints this message below two time before initializing the calculations. However, it stops when prints this message for the first time. Maybe it gives some clues to find out how get rid of this?

=========
Getting symbols from BC object
Getting cluster names from atoms object
Finished reading cluster_info
Reading basis functions from BC object
Reading translation matrix from BC
Reading translation matrix from list of dictionaries
Inserted 2688 into the translation matrix
Parsing correlation function
CEUpdater initialized sucessfully!
==============

**Comment by davidkleiven** [comment]
Yes, that's just for debugging, it will be removed in the future. I suspect that the CE code maybe will use MPI to parallelize certain aspects of the code, and it will be resolved there. So a hack now is to first run *get_ce_calc* with a single processor then **keep** the database, and the next time call it using MPI. If the database exist then all cores will just read and that is fine.

**Comment by phymalidoust** [comment]
Doesn't help. 
Why it says for the first time that "CEUpdater initialized sucessfully!" but dies for the second round? I guess get_ce_calc used two different subroutines ? Maybe we can get rid of the second round and let it to intializes the calculations with the first round?

**Comment by davidkleiven** [comment]
Do you have a minimal example that demonstrate this? What do you mean by "Maybe it gives some clues to find out how to get rid of this"?

**Comment by phymalidoust** [comment]
Here is a minima example. I meant it can be around where this error happens (attached
[mpi_test.py.zip](https://github.com/davidkleiven/CEMC/files/2755068/mpi_test.py.zip)

).

`from __future__ import print_function
import numpy as np
from mpi4py import MPI
import sys
import time
import os

comm = MPI.COMM_WORLD
num_ranks = comm.Get_size() 
my_rank = comm.Get_rank() 
print('PROCCESSOR IS:', my_rank)

ECI_FILE = "/home/mohamal/RUNs/parallel_py/data/almgzn_4x4x4.json"
db_name = "/home/mohamal/RUNs/parallel_py/data/almgzn_4x4x4.db"

folder = "/home/mohamal/RUNs/parallel_py/data/energies"
pname = str(my_rank)
fname = folder + '/' + 'processor_' + pname + '.txt'
Lg_dbname = folder + '/' + 'cpu_' + pname + '_large_4x4x4.db'


def main():


    if my_rank != 0:
        sr_comm = comm.recv(source=my_rank-1)
 

    from cemc import get_ce_calc
    from ase.clease import CEBulk
    from ase.clease import GenerateStructures
    from ase.clease import Evaluate
    import json


    conc_args = {
        "conc_ratio_min_1": [[8, 0, 0]],
        "conc_ratio_max_1": [[0, 8, 0]],
        "conc_ratio_min_2": [[6, 3, 4]],
        "conc_ratio_max_2": [[2, 5, 6]]
    }
    

        
    argms = {
    "crystalstructure":"fcc", "a":4.05, "size":[4, 4, 4],
    "basis_elements":[["Al", "Mg", "Zn"]], "conc_args":conc_args,
    "db_name":db_name, "max_cluster_size":4
    , "max_cluster_dia":[0, 0, 5.2, 4.1, 4.1]
    }

    
 
    ceBulk = CEBulk( **argms)

 
    mc_cell_size = [10, 10, 10]

    with open(ECI_FILE, 'r') as infile:
        eci = json.load(infile)

##    eci = {"c1_0": 0.1}
        

    
 
    calc = get_ce_calc(ceBulk, argms, eci=eci, size=mc_cell_size, db_name=Lg_dbname)

    ceBulk = calc.BC
    ceBulk.atoms.set_calculator( calc )


    struc_generator = GenerateStructures(ceBulk, struct_per_gen=4)

    if my_rank != num_ranks-1:
        comm.send(1, dest=my_rank+1)


if __name__ == "__main__":
    main()
`

**Comment by davidkleiven** [comment]
@phymalidoust This script is not supposed to work with MPI. What are you trying do parallalise over?

**Comment by davidkleiven** [comment]
As it is now calling MPI will just result in exactly the same instructions running on multiple processors, which is likely to cause IO problems.

**Comment by phymalidoust** [comment]
> @phymalidoust This script is not supposed to work with MPI. What are you trying do parallalise over?

The script is quite simplified. It supposed to make few databases for the large structures with multiple nodes in parallel.
By this script, the processors don't work in parallel right now. One works after the previous one is done. This ensure that there's no mixup among the processors. The first processor even produces its own database but stuck somewhere after ...

**Comment by davidkleiven** [comment]
So you planned to create many databases, one for each core? In that case why don't you just include the *rank* in the database name and all other files used for IO?

**Comment by phymalidoust** [comment]
Actually not planned to produce many dbs in actually calculations. I just narrowed down the code to isolate the issue up to 'get_ce_calc' and this simplified code shows in what function a multiprocessor computation is stopped.

**Comment by davidkleiven** [comment]
Yes, so you first has to prepare it, then run in parallel. Maybe I will change this, we don't have to that. One processor can create it, the other can wait. And then they can read. Actually, I will do that. Good that you brought this up.
